### ISSUE TITLE
style guide: switch to snake case for functions

### ISSUE THREAD

**Issue by isaachier** [ziglang/zig#1097]
I understand the general rule does not use `snake_case` for function. Why do `eql_slice_u8` and `hash_slice_u8` break this rule?

**Comment by tgschultz** [comment]
There are actually quite a few places where the std lib breaks formatting guidelines or is generally inconsistent. For example, may fields in the structs produced by `@typeInfo` are snake_case despite being of type `type`, std.mem has `readIntBE` and `readInt` while std.io has `readIntBe` and `readVar`, etc.

I think it is just a case of being pretty low on everyone's priority list until zig nears 1.0.0.

**Comment by bheads** [comment]
I would also want to standardize names (printAlloc vs allocPrint, isXXXX setXXXX getXXXX findXXXX) and common param orders (ie is alloc always the first param except when passing self). Also the stdlib should set a good example and use enums judiciously even for simple flags.

**Comment by andrewrk** [comment]
I am amending this proposal with how to handle initialisms and acronyms in TitleCase things. In summary: Don't mix-mangle the case of initialisms or acronyms.

Before:

`fn readU32Be() u32 {}`
`openZ()` `open0()`
`Os`
`Json`
`Url`
`XmlHttpRequest`
`ServeHttp`
`ServeHttpOs`

After:

`fn read_u32_be() u32 {}`
`open_z()` `open0()`
`OS`
`JSON`
`URL`
`XML_HTTP_Request` (use underscore even in TitleCase to indicate border between initialisms/acronyms)
`XML_HTTP_Request2`
`XML_HTTP_RequestRead_JSON` (no underscore between Request and Read)
`XML_HTTP_RequestRead_JSON_u32` (don't mangle primitive type names for the sake of TitleCase)
`XML_HTTP_Request_u32` (need underscore to show where Request ends and u32 begins)
`XML_HTTP_Request_u32_2`
`ServeHTTP`
`ServeHTTP_OS`

If those things were functions:

`os()`
`json()`
`url()`
`xml_http_request()`
`xml_http_request2()`
`xml_http_request_read_json()`
`xml_http_request_read_json_u32()`
`xml_http_request_u32()`
`xml_http_request_u32_2()`
`serve_http()`
`serve_http_os()`

Reasoning:

Our previous camelCase style has flaws:

 * The information it attempts to communicate does not work for single-word function names.
 * It requires case-mangling acronyms/initialisms.

This is unfortunately a widely breaking change and feels annoying to update code just for an arbitrary style decision, so my suggestion for how to roll this out, if accepted, would be to wait until we have the ability to do identifier-renaming (the feature that some IDEs provide) and it works well. That feature would ease the burden of transitioning a codebase from one style to the other, and we could even provide a script for people to run which would make the changes automatically (the script would have listed all the std lib identifiers that changed; it would not arbitrarily change users' own identifiers).

**Comment by jorangreef** [comment]
I came here to issues hoping to find there was a plan to fix the style guide.

Awesome to see this will be addressed.

**Comment by artob** [comment]
The ayes would seem to have it (good, IMHO).

@andrewrk For those of us busy building Zig libraries, it would be rather helpful to know soonest whether this proposal--per the examples you gave above--will be accepted or not, so as to not have to frustrate downstream users with massive renames later on.

**Comment by Rocknest** [comment]
@artob username mrijnyk does seem to be your alt account. Empty profile, follows only you, stars only your repos.. This fact does not change result of 'the vote' much but it does not matter since the decision should be made after considering all pros and cons, and not because of some yay/nay voting

**Comment by andrewrk** [comment]
Note to future self in about 2 weeks: don't postpone this one to 0.8.0 make a decision!!

**Comment by Rocknest** [comment]
Right now the style guide helps to see when something is a function or a variable or a namespace, with this only `type` would be easily recognisable. In IDE you would see the type kind (and much more) on hover, however it is nice to have that communicated in plaintext.

**Comment by andrewrk** [comment]
That's true, and was the main motivation for status quo style, but it doesn't work for functions with only 1 word in them, so there was always that cognitive dissonance.

BTW did you know that ZLS+vscode has semantic highlighting? It makes functions a different color, so you don't even have to hover.

**Comment by Rocknest** [comment]
I haven't tried ZLS yet. 

Before i mostly used Java which has similar preferred style to Zig and i not even once had a problem naming a function that has an acronym or an abbreviation. Also it seems that Java's preferred style is no longer all caps acronyms, for example few recently added classes `java.time.chrono.IsoChronology` and `java.net.http.HttpClient`

**Comment by tauoverpi** [comment]
While semantic syntax highlighting is nice it requires editor support and the experience for non LSP users is degraded. Basic completion methods work rather well the minute the second word is hit which would be made less effective by snake_case matching functions, variables, and so on. The shape change is, in my opinion, much easier to follow as TitleCaseTypes clearly separate from regularFunctions and variables_with_snake_case.

On the topic of semantic highlighting, switching highlighting based on what you want to focus on as in https://buttondown.email/hillelwayne/archive/syntax-highlighting-is-a-waste-of-an-information/ would (arguably) be better with a shape difference since the highlight focus changes based on the task and removing the difference between functions and other things for the other modes.

**Comment by jdknezek** [comment]
FWIW I'm a fan of `camelCase` for function/value distinction, and a fan of case mangling to avoid awkwardness of things like `XMLHTTPRFCReader`.

**Comment by SpexGuy** [comment]
Talked with @andrewrk and @thejoshwolfe.  We decided to keep approximately the current naming conventions, but clear up any ambiguous cases with the following specific algorithm:

To produce a name, first write out the name all lowercase with spaces.  Then, consider the type of variable:
- instantiated (non-namespace) type => capitalize each word and delete the spaces.
- function => capitalize each word except the first and delete the spaces.
- anything else => replace the spaces with underscores.

If at any time you would delete a space and the character to the right of the space is a number, instead replace the space with an underscore.  `anytype` values fall into the "anything else" category.

Some examples:
`xml http request` => `XmlHttpRequest`
`read u32 be` => `readU32Be`
`read u 32 id` => `readU_32Id`
`open u32 2` => `openU32_2`
`os handle` => `OsHandle`
`a b test` => `ABTest`
`ab test` => `AbTest`
`value as u32` => `value_as_u32`
`my u32 module` => `my_u32_module`

This transformation is both unambiguous and invertible.

**Comment by daurnimator** [comment]
@SpexGuy any recommendations for if the name starts with a number?

**Comment by SpexGuy** [comment]
Uh, taking off my "official" hat for a second, but maybe don't?  I don't think it's unreasonable to disallow identifiers which start with numbers in the standard library.

**Comment by daurnimator** [comment]
@SpexGuy one such example might be crypto functions; at the moment we are going to have:
```zig
pub const aead = struct {
    pub const aegis = struct {
        pub const Aegis128L = @import("crypto/aegis.zig").Aegis128L;
        pub const Aegis256 = @import("crypto/aegis.zig").Aegis256;
    };
    .....
};
```

It would make a lot of sense to have `aead.aegis.128L` instead of `aead.aegis.Aegis128L`, where we only really keep the `Aegis` prefix so that the name doesn't start with a number.

**Comment by jedisct1** [comment]
AEGIS128l is the name of the algorithm, not "128l". Splitting that name wouldn't make sense, and would break grepability as well.

**Comment by SpexGuy** [comment]
Oh, yeah, if you would capitalize a word and it starts with a number, don't capitalize it, and the space to the left of it will always become an underscore because it's a number to the right of a space.  So under these rules it would be
`aegis 128 l` => `Aegis_128L`.  That's a little weird but unambiguous.

**Comment by MasterQ32** [comment]
@SpexGuy @andrewrk should we link the answer above in the FAQ, so the current style guide is easily discoverable?

**Comment by ifreund** [comment]
> any recommendations for if the name starts with a number?

Identifiers can't start with numbers unless you use `@"123"` syntax:
```
IDENTIFIER
    <- !keyword [A-Za-z_] [A-Za-z0-9_]* skip
     / "@\"" string_char* "\""                            skip
```

**Comment by kuon** [comment]
I realize this proposal has been closed, but I started zig a while ago and I find the `camelCase` for function very disturbing.

I don't want to start a debate, but I'll just share my experience as a developer.

In short, I can read `TitleCase` and `snake_case` fine, but I have a lot of trouble with `camelCase`. I've been coding for 30 years, and its always been a pain point. In my head when I see a `camelCase` word there is a long pause, like it was written `camel. Case`. Which makes the code much harder for me to read.

Apparently, I am not alone, I found a study about this <https://ieeexplore.ieee.org/document/5521745> but I do not know how significant this is.

Again, I do not want to start a debate, nor even think of changing anything in zig. But I do use `snake_case` for functions in all my code (and `TitleCase` for types). And I hope the code I write (libraries (I have a lot of projects)..) will be accepted by the `zig` community without being frowned upon as "bad style".

**Comment by ifreund** [comment]
Seeing this reopened has given me renewed hope that the official style guide might be switched to snake_case for functions. I personally find snake_case significantly more readable than camelCase, especially for longer multi-word identifiers and at small font sizes/greater viewing distance.

If anyone would like to see a significant example of mission-critical Zig code written with snake_case functions, check out [tigerbeetle](https://github.com/tigerbeetledb/tigerbeetle). In particular, the Viewstamped Replication protocol implementation in [replica.zig](https://github.com/tigerbeetledb/tigerbeetle/blob/main/src/vsr/replica.zig) is good reading. Consider the readability of e.g `count_message_and_receive_quorum_exactly_once()` vs `countMessageAndReceiveQuorumExactlyOnce()` at different font sizes and viewing distances. Also try taking off your glasses if you wear them, if you don't use your empathy and imagination.

As pointed out above, using camelCase to distinguish function identifiers from other snake_case identifiers only works for identifiers with more than one word which makes me question the entire premise. A rule like this is a lot less valuable when all it can say about single word identifiers is "maybe it's a function, maybe not."

I'd also like to note that most if not all open source C projects I have contributed to are written with snake_case functions. Perhaps this is largely influenced by the Linux kernel style guide. In any case, switching to snake_case function names for Zig code may make working on projects using a mix of C and Zig easier.

**Comment by jedisct1** [comment]
I've never been a fan of camel case either.

Given how it's used, it's fairly obvious to understand if an identifier is a local identifier or a function, especially since Zig doesn't allow shadowing.

So, camel case doesn't improve readability, and doesn't help avoid confusion/bugs either.

**Comment by kuon** [comment]
I just stumbled on an HN article that point to as study about this.

https://whatheco.de/2013/02/16/camelcase-vs-underscores-revisited/

The conclusion is that snake_case is about 20% faster to read.

In prior studies the result was that it was a bit slower to write but if we want to emphasis on reading it seems that this may matters.

-- 

Nicolas Goy
https://www.kuon.ch




-------- Original Message --------

**Comment by andrewrk** [comment]
Anyone have any ideas how to roll out a change like this with minimal pain and suffering?

**Comment by kuon** [comment]
Write a script that can be used with other codebase too and do it in one commit per branch. The script should traverse the codebase, produce a list of identifier for manual check (this can take several iterations) and then replace them everywhere, doc included.

-- 

Nicolas Goy
https://www.kuon.ch




-------- Original Message --------

**Comment by jayschwa** [comment]
It'd be nice to have the old and new names coexist for a release cycle so that third-party code can reasonably target both the latest tagged version of Zig and development builds. As an example:

| Zig Version | `shaveYakHerd` | `shave_yak_herd` |
| --- | --- | --- |
| v0.10.0 | Present | N/A |
| v0.11.0-dev | Deprecated (Alias) | Added |
| v0.11.0 | Deprecated (Alias) | Present |
| v0.12.0-dev | Removed | Present |

I think `const shaveYakHerd = shave_yak_herd; // Deprecated in v0.11.0.` would be sufficient. But maybe there's an opportunity for something fancier like `const shaveYakHerd = std.meta.deprecated(shave_yak_herd, semver_0110)` that can `@compileLog` something helpful after a certain version is exceeded or a "log deprecations" flag is set.

**Comment by nektro** [comment]
update `zig fmt` to do the conversion, announce a quick code freeze on many channels, merge as many prs as possible, run zig fmt on the main zig repo, allow prs again

**Comment by kuon** [comment]
As zig is still in what we could call "in development" I think it is reasonable to not have a compatibility layer but provide a script that would update third party codebase in a single run.

For example, root.log did break and now it is std_options.logFn so the community is used to it.

We can provide the script and ask the community to test it to gather feedback. I think that would be easier to deal with it this way. With a clear before and after.

Deprecation is more justified when there is a rewrite involved to upgrade the code, if all there is to it is to run a script, I think it is simpler without it.

-- 

Nicolas Goy
https://www.kuon.ch

**Comment by ifreund** [comment]
Assuming this proposal applies to zig's builtin functions as well, I think those should be automatically updated by zig fmt without any user intervention (i.e. transform `@intCast()` to `@int_cast()`). I don't think zig fmt should convert identifiers from camelCase to snake_case, we've explicitly stated that such conversion is out of scope for zig fmt in the past.

I personally don't think trying to expose both camelCase and snake_case identifiers for a full release cycle is worth the effort. As @kuon says, semantics aren't changing and updating should be as simple as a find and replace. I think the best thing we can do to reduce pain is to provide good tooling to do this find and replace at scale.

To put my money where my mouth is, I am currently writing a prototype program to convert camelCase to snake_case. Here is my planned CLI, based on zig fmt:

```zig
pub const usage_camel2snake =
    \\Usage: zig camel2snake [file]...
    \\
    \\   Converts camelCase identifiers to snake_case in the input
    \\   files and modifies them in-place. Arguments can be files
    \\   or directories, which are searched recursively.
    \\
    \\   Only identifiers appearing in the whitelist and not appearing
    \\   in the blacklist are converted. By default both lists are empty
    \\   and nothing will be converted.
    \\
    \\Options:
    \\   -h, --help             Print this help and exit
    \\   --color [auto|off|on]  Enable or disable colored error messages
    \\   --exclude [file]       Exclude file or directory from processing
    \\   --whitelist fooBar     Add identifier fooBar to the whitelist
    \\   --blacklist fooBar     Add identifier fooBar to the blacklist
    \\   --whitelist-all        Whitelist all identifiers
    \\
;
```

It may also be desirable to have `--whitelist-file` and `--blacklist-file` options that read a list of newline-separated identifiers from a file. Another interesting option would be `--whitelist-std` which would whitelist a hardcoded list of camelCase symbols exposed by the standard library at the point in time that the standard library was converted to snake_case.

Regardless of how the whitelist/blacklist are constucted, the method of operation I intend to use is as follows:

1. Tokenize the target .zig files using a custom tokenizer that ignores the zig grammar and splits the input into contiguous chunks of upper/lower case ASCII plus numbers and containing at least one capital ascii character. For example the following code would be tokenized as `sliceTo, fooBar, fooBar`:
```zig
/// see also std.mem.sliceTo()
pub fn fooBar() []const u8 {
    return "fooBar";
}
```
This has the advantage of finding camelCase identifiers that appear in comments and string literals, which should also be updated.

2. Write the raw source code to the output file without doing any formatting. The only change is to convert tokens from the above tokenization that appear in the whitelist but not the blacklist to snake case. This will use the following simple algorithm:
```zig
fn camel2snake(camel: []const u8, writer: anytype) !void {
    for (camel) |c| {
        if (std.ascii.isUpper(c)) {
            try writer.writeByte('_');
            try writer.writeByte(std.ascii.toLower(c));
        } else {
            try writer.writeByte(c);
        }
    }
}
```

I'm sure there are some edge cases that will be missed by this relatively simple approach. What I don't know is how prevalent they will be and whether or not the complexity of handling them automatically will be worth it given their prevalence. I think the best thing to do is write the script and try it out to see what goes wrong and what the pain points are.

**Comment by jayschwa** [comment]
> I personally don't think trying to expose both camelCase and snake_case identifiers for a full release cycle is worth the effort.

Maybe I misunderstand the effort. I thought some `[pub] const oldFunc = new_func` lines would be a simple way to ease the migration, especially if it's a hypothetical rename tool inserting them. A one-and-done is certainly easier for the Zig project, but downstream packages will have to choose "which side" of the rename to support. I maintain the Ubuntu package of Zig, and the install base is fairly split. One third of installations follow master and two-thirds follow tagged releases.

**Comment by ifreund** [comment]
> Maybe I misunderstand the effort. I thought some [pub] const oldFunc = new_func lines would be a simple way to ease the migration, especially if it's a hypothetical rename tool inserting them. A one-and-done is certainly easier for the Zig project, but downstream packages will have to choose "which side" of the rename to support. I maintain the Ubuntu package of Zig, and the install base is fairly split. One third of installations follow master and two-thirds follow tagged releases.

My take is that adding temporary `[pub] const oldFunc = new_func` lines to the standard library as you describe isn't really going to solve the problem of zig projects needing to decide "which side" of the rename they want to support. There have already been several breaking changes since the latest release that *don't* have such compatibility attempts (such as the new std.options system) that will prevent zig 0.10 projects from building with zig's master branch/0.11 when it is released.

The main effect I expect from adding such `[pub] const oldFunc = new_func` lines to the standard library is that it will take significantly longer for the ecosystem to fully migrate to the new APIs, and that there will be a second round of massive breakage when the compatibility shims are removed.

**Comment by mlugg** [comment]
I don't have overly strong opinions on this (I like the current style but wouldn't lose sleep about switching to `snake_functions`). However, one request: if we end up implementing this and somehow reformatting all of `std` to match it, since that's a huge breaking change, can we fix up the other naming issues in `std` while we're at it? Here are a few examples:
- most of the enum alternatives in `builtin.zig` use `PascalCase` where they should use `snake_case`
- same for `std.builtin.Type`'s alternatives
- (one particular case there: renaming the `std.builtin.Mode` alternatives gives us `Mode.release_fast`, `-OReleaseFast`, and `-Drelease-fast`, which is pretty awful. Can we at least change the optimisation flag to `-Orelease-fast` to be a bit more consistent?)
- all the `type` fields in `std.builtin.Type` should use `PascalCase` (e.g. `std.builtin.Type.Pointer.child` -> `std.builtin.Type.Pointer.Child`), unless we want to amend the style to only apply `PascalCase` for decls
- `std.debug.TTY` -> `std.debug.Tty`
- `std.json.Token`, `std.json.Value`, <many more `std.json` things> alternatives should use `snake_case`

The style guide should also probably be amended to clarify the following:
- `snake_labels: {}`
- `error.PascalErrors`
- `Foo.zig` implies `Foo/` as the corresponding dir

**Comment by jayschwa** [comment]
> There have already been several breaking changes since the latest release that don't have such compatibility attempts

I don't maintain a lot of Zig code (yet), but I have found breaking changes over the last couple years relatively easy to work around in my own code. Historically, it hasn't taken me much more than a `@hasDecl`, a compiler version check, or a small shim to support both master and the latest tagged release at the same time. I don't think that strategy will work for a mass rename event.

I don't feel strongly about it, but the question of how to "minimize pain and suffering" was posed. This is just my perspective as someone that tries to target both master and release when possible.

**Comment by r00ster91** [comment]
I think after #1717, it'll make even more sense to do this because functions will be on the same level as any other declarations.
If #1717 somehow doesn't happen, we might take that as an excuse to not do change the style because "they're special".

**Comment by kuon** [comment]
@BanchouBoo I agree that style is personal and that you might prefer one or another. But I diagree with the fact that `camelCase` can be used alone to determine that an identifier is a function, because many functions are just one word. Also, in the standard library the prefix `Fn` is used to determine that it is a function (`std_options.logFn` for example). Which would become `log_fn`. That prefix is much better to have a clear information that the variable is a function than playing with case. Also, this kind of prefix works fine with no syntax highlighting.

For me, the main argument against `camelCase` is two-folds beside personal preference:

- as I said above, `camelCase` alone is not enough to identify a function, thus it is just a matter of consistency with zig itself as we already use `snake_case` for  identifiers
- the second one is more a feeling than a fact, but my feeling is that zig is getting momentum from programmers that are more used to `snake_case`, like linux C programmers, rust programmers and also erlang/elixir programmers writing NIF. 

This being said, I feel strongly about freedom of style, and as I have me preference, @BanchouBoo  has others, so as I asked in in #14228  zig should keep accepting any style like C does, so this part is covered. For me this proposal is good, but as it has been said many times, there is not "super strong argument" in either side. My feeling is (and I am biased because I like it) that it would benefit the community and the adoption of `zig`.

Now, if we want to go through with this proposal, I think it is really important to make it a "one time breaking change with a very straightforward upgrade path".

So to summarize what has been said:

- write a tool (`zig fmt` ?) for the upgrade
- upgrade all branches at once in this repo in the fewest commits possible, block any PR while doing so, postponed PR will need to have the tool ran on them
- having a compatibility layer in the form of `const fooBar = foo_bar`, while tempting, would not really help with the upgrade and would just make the transition much slower and would introduce some kind of technical debt, we don't want a python 2-3 situation
- as @mlugg said, the std lib has some inconsistencies, they can be included in the upgrade only if the tool can automatically take care of it, this would require a seperate list, manually curated or words with their replacements

**Comment by silversquirl** [comment]
I like `camelCase`. I prefer it to `snake_case` in general because I find it faster to write and, more importantly, easier to read (for the "it feels like there's a pause" reason mentioned by @BanchouBoo and, in the inverse, by @kuon).

However, there are some issues with `camelCase` that aren't present with `snake_case`:

- Acronyms. Neither `decodeXmlRfcSpec` nor `decodeXMLRFCSpec` is a good option
- Numbers. You can't capitalize them, which results in names being more ambiguous.
  It's particularly difficult in cases where you have multiple numbers in sequence, eg. a function that reads RGB using 10 bits for red, 12 bits for green and 10 bits for blue - the snake_case name is obvious (`read_rgb_10_12_10`) but the camelCase name less so (`readRgb101210`? Hard to read. `readRgb10_12_10`? Not really snake case any more)
- Single-letter abbreviations, like `u` for unsigned. If you have a function for reading signed integers you might call it `readInt`. The unsigned equivalent would be `readUint`, but now the case of the `I` in `Int` has changed and it's less obviously a variant of the same function. You could also use `readUInt` but that looks very strange. `read_int` and `read_uint` is more consistent and more visually similar.

Because of those issues, I think `snake_case` is a better overall, despite me personally preferring `camelCase` for simple cases. This consistency and readability in more obscure situations is particularly important to Zig's philosophy.
As written in the Zen, edge cases matter.

**Comment by SuperAuguste** [comment]
I think the next logical step would be changing type naming from `ImpossibleToReadFromALongDistanceTitleCase` to `This_Is_Easier_To_Read` or even just normal snake case. The latter is also what C does.

**Comment by kuon** [comment]
@SuperAuguste I agree that `TitleCase` it not super easy to read, but it also helps to ensure type names aren't too long. Also they can be namespaced. I also think doing the same than `rust` would help language adoption.

**Comment by mlugg** [comment]
To me, `PascalCase` takes no longer to visually parse than `snake_case` (while `camelCase` is noticeably worse). Moreover, differentiating type names has some concrete advantages:
- it allows easily differentiating a type and its corresponding `std.builtin.Type` when they're both assigned to names
- it gives us obvious names for initialization functions for certain generic types (e.g. `BufferedReader` vs `bufferedReader` (to become `buffered_reader`))
- it makes the usage of a type obvious without having to read its definition (we know that `array_list` is a namespace while `ArrayList` is an instantiatable type)

**Comment by BanchouBoo** [comment]
> But I diagree with the fact that camelCase can be used alone to determine that an identifier is a function, because many functions are just one word.

I addressed this in my post, but to reiterate, I don't think the presence of some ambiguity is a good reason to make it completely ambiguous. That doesn't solve the problem.

> Also, in the standard library the prefix Fn is used to determine that it is a function (std_options.logFn for example). Which would become log_fn. That prefix is much better to have a clear information that the variable is a function than playing with case. Also, this kind of prefix works fine with no syntax highlighting.

This is about more than just a few specific things in the standard library that will be clear regardless though, this is about the accepted and encouraged style for all code. I do think prefixes and suffixes could potentially be a good way to disambiguate regardless of the style chosen, but I also think it feels wrong to try and enforce more specific parts of what a name should be like that as convention, and I'm not sure what the best approach would be in that context. Appending or prepending functions with `fn`/`Fn` feels awkward, perhaps making functions PascalCase and then suffixing types with T, so for example `const MyStructT = struct{}`.

Alternatively, but probably more controversially and unlikely to be accepted would be to have some unique syntax for referencing a function. This feels pretty clunky to me as a solution though.

> the second one is more a feeling than a fact, but my feeling is that zig is getting momentum from programmers that are more used to snake_case, like linux C programmers, rust programmers and also erlang/elixir programmers writing NIF.

C code style is pretty inconsistent between codebases, though, doesn't seem like a great point of reference. I don't think that the encouraged code style is going to have a significant effect on adoption one way or another. I very much doubt anyone would see the change to snake_case and say "finally, time to try zig!"

> Because of those issues, I think snake_case is better overall, despite me personally preferring camelCase for simple cases. This consistency and readability in more obscure situations is particularly important to Zig's philosophy.

These are definitely problems I've encountered numerous times, I'm never quite sure how I should name things in those cases. But while it's easier in snake_case to not have to worry about those situations, it's not like it's impossible to define style guide rules for those contexts either. Ultimately I value keeping things unambiguous between functions and variables above everything else.

**Comment by silversquirl** [comment]
> it's not like it's impossible to define style guide rules for those contexts either

Sure, but there are no good options. All the reasonable options for those edge cases have readability or consistency issues.
Also, adding more special cases isn't really a good solution. It's better to have a naming scheme that's consistent, obvious and readable for all cases, which snake_case is.

**Comment by sanks64** [comment]
> I think the next logical step would be changing type naming from ImpossibleToReadFromALongDistancePascalCase to This_Is_Easier_To_Read, THIS_IS_EASIER_TO_READ, or even this_is_easier_to_read so typefuncs fit in with other functions more evenly. The latter is generally what C code does.

What I never see mentioned is that `This_Is_Easier_To_Read` has advantages:

1. Easy to convert type names to function names:

   - `XMLParser` is `xml_parser` or `x_m_l_parser`?
   - `XML_Parser` is `xml_parser`.

   Of course, in the current situation, you would just use `XmlParser`, but in English you write `XML` and not `Xml`, so it feels like a hack.

2. It aligns with `snake_case` better:

   - Old for types and old for functions:
     ```c
     // Easy to convert
     // Hard to read (the only difference is a single character)
     // Easy to type (no need for an extra underscore)
     XmlParser
     xmlParser
     ```
   - Old for types and new for functions:
     ```c
     // Hard to convert
     // Hard to read (different widths)
     // Easy to type (no need for an extra underscore)
     XMLParser
     xml_parser
     ```
   - New for types and new for functions:
     ```c
     // Easy to convert
     // Easy to read (same widths)
     // Hard to type (needs an extra underscore)
     XML_Parser
     xml_parser
     ```

So I'd advocate for at least considering snake Pascal case (or whatever it is called).

**Comment by castholm** [comment]
Re: `PascalCase` vs `Pascal_Snake_Case`: If having simple rules with no edge cases is a goal it might be worth pointing out the algorithm for the latter is simpler:

* `PascalCase`
  1. Write out the name in lowercase with spaces.
  2. Capitalize each word.
  3. If the character before a space is a number, replace the space with an underscore; else, remove the space.
* `Pascal_Snake_Case`
  1. Write out the name in lowercase with spaces.
  2. Capitalize each word.
  3. Replace each space with an underscore.

Also worth mentioning is that `PascalCase` has the same readability/consistency problems with short acronyms (`BufferedIoStream`), numbers (`Rgb3_3_2`) and variants (`IntParser` -> `UIntParser`? `UintParser`?) as `camelCase`. Always using underscores as separators helps alleviate those to some extent. (But I'm not sure if they are really that much of a problem in practice.)

For what it's worth I personally like the current (as of 0.10.1) recommended convention a lot and I think there is utility in having identifiers of types, functions and values take on a different visual appearance when written out (without littering code with Hungarian notation and prefixes/suffixes like `fn` or `t`). It makes source code slightly easier to visually parse in situations where semantic highlighting isn't an option. But people have already brought up those points plenty and whichever way things end up the most useful thing in the end is to just have a suggested convention the ecosystem can fall back.

**Comment by kuon** [comment]
I agree with what @sanks64 said, but I still think `TitleCase` for type is OK because it:

- is the same/compatible with `rust`, yes this is not a competition, but anything that can make adoption smoother is a win, even something as tiny as this
- is not a problem with long names, because in your example, I would do `xml.Parser` in a namespace, `http.Request`...
- is doing its job with disambiguation, starting in uppercase -> type
- can be aliased, if you do not want to use the namespace AND it would be ambiguous to just use `Parser`, you can do `const XML_Parser = xml.Parser` and not use `xml.Parser` . You got freedom.

Last (this part is highly personal), for me name in `TitleCase` have slightly different "problematic" than `camelCase` for functions. Functions with long descriptive name are a thing, like `open_file_seek_and_close`, it is a reality of real world applications. They can also represent a sequence like my example, thus the need for better separation of words. But types can often be shorter, and if they are two/three words, those words are one "unit", like `FixedBufferAllocator`, it is one logical unit. I do not actually "read" the text, I "see" the text as a single "blob".

**Comment by IntegratedQuantum** [comment]
> I just stumbled on an HN article that point to as study about this.
> https://whatheco.de/2013/02/16/camelcase-vs-underscores-revisited/
> The conclusion is that snake_case is about 20% faster to read.

On first glance I thought "Although I prefer camelCase, snake_case must be better, science proves it".
But then I dug deeper and found that the same authors 2 years later did a follow-up study:
[The impact of identifier style on effort and comprehension (2012)](http://www.cs.kent.edu/~jmaletic/papers/EMSE12.pdf)
There they tried a couple of different experiments with more mixed results:
One experiment favors camelCase, one favors snake_case and the others don't show a significant difference between camelCase and snake_case.

**Comment by kuon** [comment]
@IntegratedQuantum yeah as I said https://github.com/ziglang/zig/issues/1097#issuecomment-1405588433 , science is not a strong argument in either direction. For me the familiarity of either style is stronger.

**Comment by sanks64** [comment]
> - is the same/compatible with rust, yes this is not a competition, but anything that can make adoption smoother is a win, even something as tiny as this

@kuon I agree that it makes adoption smoother, but I think in this case being better is more important than being familiar.

If talking about preference, I would go with `XMLParser` or `XML_Parser` because `Xml` irks me each time I read it.

Lastly, I agree that having `snake_case` for functions is more important.

**Comment by kuon** [comment]
@sanks64 for this particular example, if we do not namespace, I agree, `XMLParser` or `XML_Parser` is better, but I would still use a namespace personally.

**Comment by sanks64** [comment]
@kuon Yep, I agree that using a namespace is a good suggestion, but we can't rely on it in all situations.

**Comment by kuon** [comment]
> @kuon Yep, I agree that using a namespace is a good suggestion, but we can't rely on it in all situations.

Yes it depends. For example, I would not write `Hash_Map` because for me it is really a "single word", a bit like the engligh words grandfather, parking lot, afternoon... But the `XML` example would be perfectly fine as `XML_Parser` if a namespace cannot be used for any reason, which as you said, is possible.

I think it is reasonable to keep `TitleCase` while not being super strict about it, a few exceptions that would not shock me, on the top of my mind:

- abbreviations: `Rgb` -> `RGB`, `Hsluv` -> `HSLuv`
- C names`GlFloat` -> `GLfloat` (woud clearly use a namespace here, but might not be an option)
- variants `ImageSD` `ImageHD`, maybe in autogenerated code like`Foo_1` `Foo_2`...
- common names `FrameBuffer` -> `Framebuffer`

**Comment by Deecellar** [comment]
> > @kuon Yep, I agree that using a namespace is a good suggestion, but we can't rely on it in all situations.
> 
> Yes it might depend on the situation, and freedom is important. I would not write `Hash_Map` because for me it is really a "single word", a bit like the engligh words grandfather, parking lot, afternoon... But the `XML` example would be perfectly fine as `XML_Parser` if a namespace cannot be used for any reason, which as you said, is possible.
> 
> I think it is reasonable to keep `TitleCase` while not being super strict about it, a few exceptions that would not shock me, on the top of my mind:
> 
> * abbreviations: `Rgb` -> `RGB`, `Hsluv` -> `HSLuv`
> * C names`GlFloat` -> `GLfloat` (woud clearly use a namespace here, but might not be an option)
> * variants `ImageSD` `ImageHD`, maybe in autogenerated code like`Foo_1` `Foo_2`...
> * common names `FrameBuffer` -> `Framebuffer`

I do prefer after a bit of time the: Rgb and the Xml instead of all caps, is less distracting reading

**Comment by 100phlecs** [comment]
> To put my money where my mouth is, I am currently writing a prototype program to convert camelCase to snake_case. Here is my planned CLI, based on zig fmt

Thanks for explaining your design, it's well thought out. Looking forward to this :-)
Would be cool to have zig in snake_case.

> Tokenize the target .zig files using a custom tokenizer

I think you'd be able to solve this with regex if you find the tokenizer part not working out.
If the intention is to have zero dependencies to integrate into zig then your approach makes sense.
Though I think zig stdlib would benefit from a thin Regex library wrapper over https://github.com/PCRE2Project/pcre2

**Comment by WhiteHexagon** [comment]
I am sad to have stumbled over this proposal :(  camelCase served Java well for 25+ years, and is one reason I decided to learn Zig over Rust.  'being easier to read' just depends on your background experience, but almost falls into the religious debate of 'one true brace'.  I respect that Andrew made a call on these topics a long time ago, and dont like the idea of a community that can hustle in and create such divergence from the original concepts.  Please fork and call it Z_I_G.  Same goes for this push to replace '_' with a keyword, if you dont like the existing syntax, please pick another language.

**Comment by mlugg** [comment]
> I am sad to have stumbled over this proposal :( camelCase served Java well for 25+ years, and is one reason I decided to learn Zig over Rust. 'being easier to read' just depends on your background experience, but almost falls into the religious debate of 'one true brace'. I respect that Andrew made a call on these topics a long time ago, and dont like the idea of a community that can hustle in and create such divergence from the original concepts. Please fork and call it Z_I_G. Same goes for this push to replace '_' with a keyword, if you dont like the existing syntax, please pick another language.

This is an extremely counterproductive perspective.

The point of a programming language is not to never change from its original design; if that were the case, Zig could be a very different language to how it looks today. Every single issue tagged "accepted" that wasn't created by Andrew (there are over 200) is a change from his "original concepts" which was deemed an improvement. While Andrew has the last call on what makes it in, he's also human, and having other people propose ways to improve Zig - even if just as discussion points - is a crucial part of the language's evolution.

Zig isn't designed how it is because one person says so, it's because it's what that person believes to be best for users of the language. Proposing changes to the language with reasonable justifications is an important part of how that design evolves. I don't necessarily agree with everything being proposed here, but it's important to keep an open mind and understand why people might want the language to be changed in the way they do.

If one of the main reasons you like Zig is the fact that it uses `camelCase` for functions, and if you find it this concerning that people are proposing small syntax or style changes, I'd encourage you to rethink your priorities a bit.

**Comment by fearedbliss** [comment]
I'm in favor of this change to replace all function names to snake case. It's just easier to read/distinguish with the additional separation between the characters. I'm also in favor of delaying this breaking change till before 1.0. No rush.

**Comment by cdecompilador** [comment]
agree, i support snake_case or go all out with CamelCase since those two are the main two styles followed by many c programmers, mixing 3 styles is bad idea

What i don't agree is with delaying this much more, since the refactor won't be automatic since sometimes what seems good in camelCase doesn't with snake_case, like for example getters, which usually are `name` instead of `getName`

**Comment by RaphGL** [comment]
If you allow me to bikeshed a bit here, here's my take on it :)

One of Zig's primary goals is to interop with C and be another system's programmming language. As such I think adopting conventions that are compatible with those would be beneficial since snake_case is the de facto standard for C, C++, Rust and their standard libraries (even though some C/C++ libraries have opted to use another style).

I personally would propose the following style:
- Every type and every function that returns a type is PascalCase
- Constants can use SCREAMING_SNAKE_CASE if they're just a value known at comptime 
- Enums and Errors also use PascalCase as they're not really plain constants anymore like in C and you can't just access their value without @intFromEnum
- Everything else is snake_case

```zig
// Functions
fn LinkedList(comptime T: type) type;
fn remove_duplicates() T;

// Variables
const MAGIC_NUMBER = 0x4e93fd;
const immutable_value = get_value();
var mutable_value = get_value();

// Enums and Unions
const MyEnum = enum {
    OptionOne,
    OptionTwo,
};

// becomes pascal case cause you have to unwrap it like an enum
const TaggedUnion = union(enum) {
    FirstValue: i32,
    SecondValue: f32,
};


// just plain old unions so no PascalCase
const RegularUnion = union {
    first_value: i32,
    second_value: f32,
};
```

**Comment by andrebrait** [comment]
> ```zig
> // becomes pascal case cause you have to unwrap it like an enum
> const TaggedUnion = union(enum) {
>     FirstValue: i32,
>     SecondValue: f32,
> };
> ```

Enums are constante. I'd rather see them use SCREAMING_SNAKE_CASE instead.

**Comment by andrebrait** [comment]
@RaphGL also, constants should use SCREAMING_SNAKE_CASE no matter whether they're known at compile time or not.

Outside code shouldn't know what your implementation is like and shouldn't have to change if you change your implementation either. That's also likely a good thing for your own internal code too.

I'd hate to review an MR with 200 changed lines where the only actual change was making COOL_CONSTANT into cool_constant because it was a literal but now it's taken from a function or something. And that's just for stuff you *don't export*. For stuff you do, that'd be an instant compile error for anything that uses your library and, except for specific cases, external code really shouldn't care where you took the value from.

A constant is a constant.

**Comment by andrebrait** [comment]
The problem with having different styles for different classifications of things is that it becomes very unstable under refactoring and whatnot too.

Made a variable inside a Union constant? Change its casing, now change every place that uses it, etc. now remember you broke every external thing that touches that variable (now a constant) because of that.

Or don't, and now you have an inconsistent style and it doesn't make things obvious anymore.

**Comment by RaphGL** [comment]
> @RaphGL also, constants should use SCREAMING_SNAKE_CASE (or snake_case, whatever, _a single style_ is what I mean) no matter whether they're known at compile time or not.

Yeah but zig makes no distinction between constant and immutable value and most of the learning resources recommend making things immutable whenever possible so using SCREAMING_SNAKE_CASE would litter your code with capitalized variables. 

> I'd hate to review an MR with 200 changed lines where the only actual change was making COOL_CONSTANT into cool_constant because it was a literal but now it's taken from a function or something. 

Yeah that might be a possibility if it were a comptime function. This is also only an annoyance because ZLS does not support bulk renaming yet and for people that don't use an LSP (in the future).

> And that's just for stuff you don't export. For stuff you do, that'd be an instant compile error for anything that uses your library and, except for specific cases, external code really shouldn't care where you took the value from.

If you're exporting it, people don't care how that constant is built just that it's one, so making it screaming case is ok.

___ 

I only made this distinction for constants because seeing screaming case everywhere would be too noisy. I'd be fine with that distinction personally but your criticism is very valid, I'm not sure how often this would be an issue if you just make all user facing constants SCREAMING_CASE. Feel free to point out more flaws in my suggestion :)

**Comment by N00byEdge** [comment]
I think the screaming vs snake case thing is getting a little bit off topic (but I'll pitch in that I prefer snake case for all the constants, enum variants etc). This issue is about function names.

**Comment by kuon** [comment]
It seems there some consensus on renaming functions to `snake_case`. For personal preferences, similarity to rust and others (elixir, ruby, many C libs...), and the "pain point" of mixing styles. I "think" we can consider this decided.

For constants, I think further discussions is necessary, especially when we consider zig notion of constant to be different than it is in other languages because of `comptime`, but I suggest we start a new issue, as this one is rather long and "should" be focused on function names.

**Comment by cdecompilador** [comment]
Ok, to sum up, what could be said that most of us agree to
* functions **snake_case** ✅
* types/type generators **PascalCase**, **mayus acronyms** and allow **Pascal_Snake_Case+* for some cases like acronym + name (ex: XML_Request) ✅
* constants style require a new issue 🆕
* I've seen some mention of enum variants being **PascalCase** but that should discussed a new issue 🆕

Let's close this and start discussion over how to do it, which in my opinion there is no painless way

**Comment by sanks02** [comment]
@cdecompilador I don't think this issue needs any further summaries, @kuon did it well. Types in particular are out of scope, and I don't think there is agreement on `XML_Request`, so I think everyone (including me) should strongly consider creating a new issue if we want to move a proposal with a style for types forward.

**Comment by lucianboboc** [comment]
Snake case for functions is so unproductive, function names are longer than variable names.
Zig is also more readable because functions are pascal case, just trying to copy Rust will make the language worse 😅

**Comment by softprops** [comment]
Has anyone guesstimated the breaking change impact on std lib if it itself were to reflect the proposed changes?

I think it’s standard practice to seek inspiration from std libraries when trying to understand what good looks like.

left unchanged it would be weird to have a recommendation that the std lib itself doesn’t follow.

**Comment by silversquirl** [comment]
Guesstimated `std` breaking changes: all of it :)

**Comment by mlugg** [comment]
@andrewrk

> Anyone have any ideas how to roll out a change like this with minimal pain and suffering?

I just threw together a little script which automatically converts all `camelCase` (but not `PascalCase`) identifiers in a Zig file to `snake_case`, aside from any which is an `export`/`extern`. It'll query the user for any where the conversion isn't immediately obvious (e.g. `getPath2 -> get_path_2` but `isUtf8 -> is_utf8`). We'd have to be careful applying it over something as big as the standard library or compiler, but I'd think it should work well in most codebases.

**Comment by softprops** [comment]
It would be a nice feature of `zig fmt` to aide in such a code reformatting formatting. The rust community tooling also has a  [fix](https://doc.rust-lang.org/cargo/commands/cargo-fix.html) command for these kinds of changes. A `zig fix` does not yet exist but would also be nice.

**Comment by WhiteHexagon** [comment]
I think that after 5 years this issue should be closed.  It is hard enough already getting Zig code samples out of the various LLMs / GPTs, and such a fundamental change this late in the project is going to set us back a long way.  
Not to mention all the tutorials, examples and SO QA type sites that wont get updated, and for what benefit? to try to copy rust?  
If we are trying to create a better 'C', shouldnt that mean even more stability than what 'C' had?  Isn't that why 'C' has endured all this time, and is still a 'goto' choice for many projects.  Imagine picking up a KnR C style book for Zig that would still be valid 10 years from now... yeah, I know, old man dreaming :)

**Comment by mlugg** [comment]
Optimizing Zig to be a language LLMs can write effectively is quite possibly our single lowest priority. Zig is a language written for humans; not for a glorified autocompletion engine. To be frank, if one is seriously trying to get ChatGPT or the like to output good Zig code, naming conventions will be the least of their concerns.

I'm not sure what comment gave you the impression that this proposal exists "to try to copy Rust". The proposal exists because a lot of people believe this change would improve readability of Zig code and/or simplify the style guide.

Being in v0, Zig is currently an unstable language. Once it reaches v1, the intention is that it becomes highly stable, both technically and in terms of conventions; but for now, major changes to the language, style guide, conventions, etc are all perfectly allowed, and even encouraged if they are clear improvements.

**Comment by andrewrk** [comment]
Closing again. I toyed with this in https://github.com/andrewrk/autodoc and found it to tend to cause unfortunate shadowing that was not really the case before. `camelCase` for function names I think is serving us fairly well.

That said, it's just what we're going to do in the standard library - individual projects can obviously do their own naming conventions. And sometimes projects doing their own naming conventions can be handy in its own way, the human understands that one area of code has this vibe, another has that one.

I'll also consider changing the naming conventions of builtin functions. For whatever reason, camelCase in builtin functions just feels wrong. I often mistype `@typeInfo` and `@TypeOf` in particular.

**Comment by McSinyx** [comment]
> That said, it's just what we're going to do in the standard library - individual projects can obviously do their own naming conventions.

I suppose with comptime it's possible to convert function and constant names (but not globals) from one convention to another when `@import`.  Is it possible to this without messing with lazy compilation?

**Comment by kuon** [comment]
While I understand @andrewrk choice, I think that if there are naming conflict in the std lib that are prevented only by different case it can be a little bit confusing and feels "brittle", and maybe it should be addressed in another issue.

> 
> That said, it's just what we're going to do in the standard library - individual projects can obviously do their own naming conventions. 

At any rate, it is the only important thing to me. I use snake case in JS and I have no problem with std lib and DOM lib being camelCase. As long as zig makes no technical decision on the name, it is OK.

I know I did push a bit hard for this to be implemented, and I am sorry if it was too much. In the end it is only a personal preference. Cheers.

**Comment by pyandras** [comment]
> While I understand @andrewrk choice, I think that if there are naming conflict in the std lib that are prevented only by different case it can be a little bit confusing and feels "brittle", and maybe it should be addressed in another issue.
> 
> > That said, it's just what we're going to do in the standard library - individual projects can obviously do their own naming conventions.
> 
> At any rate, it is the only important thing to me. I use snake case in JS and I have no problem with std lib and DOM lib being camelCase. As long as zig makes no technical decision on the name, it is OK.
> 
> I know I did push a bit hard for this to be implemented, and I am sorry if it was too much. In the end it is only a personal preference. Cheers.

It is a preference but beyond personal as the comments and reactions in this thread show. I fail to see the current resolution being consistent with it all or your summary/meta-analysis.

**Comment by okvik** [comment]
> Closing again.

My disappointment is immeasurable, and my day is ruined.
